### PR TITLE
Polyglot

### PR DESCRIPTION
**Pull request by ralsina**
Attempt to make the codebase py2/py3 compatible

- [x] Support installing in py2 and py3
- [x] Make tests run with py2 and py3 in travis
- [x] Keep tests passing in py2

And then progressively fix tests in py3 until they all pass (or close enough)

### CODE REVIEW

**Comment by ralsina** [comment]
This breaks math (test_slides and test_issue_363)

**Comment by ralsina** [comment]
Woohoo, 144 failures.

**Comment by ralsina** [comment]
142!

**Comment by ralsina** [comment]
140!

**Comment by ralsina** [comment]
111!

**Comment by ralsina** [comment]
Extensions are broken in py3, which is unsurprising since they do manual, obscure __import__ things, relevant tests are test_extensions and test_fancytitles

**Comment by ralsina** [comment]
110!

**Comment by ralsina** [comment]
They are going to keep failing for a while, I am done for today ;-)

**Comment by ralsina** [comment]
109!

**Comment by ralsina** [comment]
105!

**Comment by ralsina** [comment]
101! (with svglib instead of svg2rlg)

**Comment by ralsina** [comment]
100!

**Comment by ralsina** [comment]
With python 3.7 literally 1 (one) ONE test passes, compared to a couple hundred with 3.6 so ... that needs looking at :-)

**Comment by ralsina** [comment]
Looks like a bug in adjustUnits

**Comment by ralsina** [comment]
In python 2.7:

```
>>> re.split(r'(-?[0-9.]*)', '2cm')
['', '2', 'cm']
```

In python 3.7:

```
>>> re.split(r'(-?[0-9.]*)', '2cm')
['', '2', '', '', 'c', '', 'm', '', '']
```

In python 3.6: same as python 2.7

**Comment by ralsina** [comment]
Back to 100 failures.

**Comment by ralsina** [comment]
95!

**Comment by ralsina** [comment]
I think most of the remaining failures are because I changed how the target reference ids are constructed, which should be easy to fix all in one swoop.

**Comment by ralsina** [comment]
I probably broke it in 42239d3caeed3e3a786d270fa53100ddf5eec53e

**Comment by ralsina** [comment]
The worst part is that links include object ID, which means they change on each run :-(

**Comment by ralsina** [comment]
55!

**Comment by ralsina** [comment]
And that is it for today.

**Comment by ralsina** [comment]
I *think* the next problem is in syntax highlighting.

**Comment by ralsina** [comment]
Interesting one! stylesheet paths are in different order in py2 and py3:

```diff
< [INFO] styles.py:113 StylePath:['/Users/ralsina/code/rst2pdf/rst2pdf/styles', '/Users/ralsina/.rst2pdf/styles', '.']
---
> [INFO] styles.py:113 StylePath:['/Users/ralsina/.rst2pdf/styles', '/Users/ralsina/code/rst2pdf/rst2pdf/styles', '.']
```

**Comment by ralsina** [comment]
Still 55, but hey, that was just wrong.

**Comment by ralsina** [comment]
The basic difference seems to be that in python2 RL is inserting colors (specifically 0 0 0 rg) that it doesn't in python3.

The outputs are identical in many tests except for that.

**Comment by ralsina** [comment]
44!

**Comment by ralsina** [comment]
19!

And that's probably it for today.

**Comment by akrabat** [comment]
I'm loving seeing this work in progress @ralsina!

**Comment by ralsina** [comment]
@akrabat a lot of the hard stuff was already done, and now it's mostly garbage collecting (build, look at two PDFs, they are the same, mark as good) ... there are probably about 8 actual bugs left :-)

**Comment by ralsina** [comment]
9!

So, what's left are things where my local computer doesn't generate the same hashes as travis. So I will turn on gitlab ci and look at the generated artifacts there to see how far that goes.

**Comment by ralsina** [comment]
Ouch, 30 failures in Gitlab, including some tests throwing exceptions, so somewhat of a setback.

**Comment by ralsina** [comment]
This added a failure in GH and removed 10 in GL, so, mmmmkay.

**Comment by ralsina** [comment]
9/19 now: slightly better

**Comment by ralsina** [comment]
issue 146 is a real bug! It's failing to embed garuda.ttf, which is provided.

**Comment by ralsina** [comment]
7/19 !

**Comment by ralsina** [comment]
7/18

**Comment by ralsina** [comment]
Eek, test_fancytitles is not repeatable on Python 3 for some reason. On every run it gives a different hash.

**Comment by ralsina** [comment]
7/17

**Comment by ralsina** [comment]
6/13

There are a number of unstable tests in Gitlab, which complicates things.

**Comment by ralsina** [comment]
In fact, re-running the tests in GitLab changes how many fail, which is a new thing :-(

**Comment by ralsina** [comment]
1 (one) failure left in travis!
11 left in gitlab-ci!

But... that's it for today :-)

**Comment by ralsina** [comment]
Noticed GL-CI was using python 3.5... that is the last python 3 version that has random-ordered dictionaries, which could be causing our trouble with broken invariancy. Updated GL config to 3.6 and if that's good can mark that as required minimum python 3 and all will be ok.

**Comment by ralsina** [comment]
@akrabat if you can do your magic and check the artifact for the one test failing in Travis, we can probably call this done.

**Comment by akrabat** [comment]
Will do

File: rst2pdf/tests/md5/test_fancytitles.json [hunk 1]
```diff
@@ -10,6 +10,7 @@ good_md5 = [
         '3ddba1b9358cf7972e8eb0d52372bac2',
         '5386e164e5a952f88980a1089cf92fee',
         '651349f0143b26986707ef8a374a4d2b',
+        '6645b78157c295d50dad176d8e31324e',
```
Comment (ralsina) [inline rst2pdf/tests/md5/test_fancytitles.json]: Wooohooo

Comment (akrabat) [reply rst2pdf/tests/md5/test_fancytitles.json]: > Wooohooo

I can read the code changes now :)

**Comment by akrabat** [comment]
Rebased against master after merging #755

**Comment by oz123** [comment]
This is most exciting.

**Comment by ralsina** [comment]
I am very happy. This branch was **a lot** of work :)

**Pull request merged**